clients/src/main/java/org/apache/kafka/clients/consumer/internals/ListOffsetsRequestManager.java

```diff
@@ -0,0 +1,346 @@
+
+        return listOffsetsRequestState.globalResult;
+    }
+
+    /**
+     * Generate requests for partitions with known leaders. Update the listOffsetsRequestState by adding
+     * partitions with unknown leader to the listOffsetsRequestState.remainingToSearch
+     */
+    private void fetchOffsetsByTimes(final Map<TopicPartition, Long> timestampsToSearch,
+                                     final boolean requireTimestamps,
+                                     final ListOffsetsRequestState listOffsetsRequestState) {
+        try {
+            List<NetworkClientDelegate.UnsentRequest> unsentRequests =
+                    sendListOffsetsRequests(timestampsToSearch, requireTimestamps, listOffsetsRequestState);
+            requestsToSend.addAll(unsentRequests);
+        } catch (StaleMetadataException e) {
+            requestsToRetry.add(listOffsetsRequestState);
+        }
+    }
+
+    @Override
+    public void onUpdate(ClusterResource clusterResource) {
+        // Retry requests that were awaiting a metadata update. Process a copy of the list to
+        // avoid errors, given that the list of requestsToRetry may be modified from the
+        // fetchOffsetsByTimes call if any of the requests being retried fails
+        List<ListOffsetsRequestState> requestsToProcess = new ArrayList<>(requestsToRetry);
+        requestsToRetry.clear();
+        requestsToProcess.forEach(requestState -> fetchOffsetsByTimes(requestState.remainingToSearch,
+                requestState.requireTimestamps,
+                requestState));
+
+    }
+
+    /**
+     * Search the offsets by target times for the specified partitions.
+     *
+     * @param timestampsToSearch the mapping between partitions and target time
+     * @param requireTimestamps  true if we should fail with an UnsupportedVersionException if the broker does
+     *                           not support fetching precise timestamps for offsets
+     * @return A list of
+     * {@link org.apache.kafka.clients.consumer.internals.NetworkClientDelegate.UnsentRequest}
+     * that can be polled to obtain the corresponding timestamps and offsets.
+     */
+    private List<NetworkClientDelegate.UnsentRequest> sendListOffsetsRequests(
+            final Map<TopicPartition, Long> timestampsToSearch,
+            final boolean requireTimestamps,
+            final ListOffsetsRequestState listOffsetsRequestState) {
+        Map<Node, Map<TopicPartition, ListOffsetsRequestData.ListOffsetsPartition>> timestampsToSearchByNode =
+                groupListOffsetRequests(timestampsToSearch, listOffsetsRequestState);
+        if (timestampsToSearchByNode.isEmpty()) {
+            throw new StaleMetadataException();
+        }
+
+        final List<NetworkClientDelegate.UnsentRequest> unsentRequests = new ArrayList<>();
+        MultiNodeRequest multiNodeRequest = new MultiNodeRequest(timestampsToSearchByNode.size());
+        addMultiNodeRequest(listOffsetsRequestState, multiNodeRequest);
```

Comment:
it sounds like this is chaining the callback, can we add a method in MultiNodeRequest that do so? I feel it is more organic to do something like: multiNodeRequest.onComplete(callback);

Reply:
Agree, will explore that and adjust it.